### PR TITLE
feat: onUploadBegin prop

### PR DESCRIPTION
**Pull request by Mr0Bread**
Addresses #109 

Adds new prop `onUploadBegin` to components. Prop accepts the callback. Callback receives single param - name of the file upload of which have just started. The callback is called once `loadstart` event of XHR fires

P.S. Will add same stuff to SolidJS package if such approach is approved

### CODE REVIEW

**Review by juliusmarminge — COMMENTED** [review]
Is there any difference between this and onUploadProgress and checking if progress is 0?

**Comment by markflorkowski** [comment]
I'll second what Julius is asking, but am also ok with the approach of adding an additional optional callback for this case as it is a bit more readable if you are not using progress, and only want some action on uploads starting.

**Comment by Mr0Bread** [comment]
@juliusmarminge @markflorkowski 

Firstly, I noticed that unUploadProgress not always being called with 0. I guess it happens when file size is small. Secondly, there is delay between call of unUploadBegin and first invocation of onUploadProgress. The amount of delay can be seen in the attached screenshot in second log
image
Here is the code that generates such an output
```jsx
onClientUploadComplete={(res) => {
  // Do something with the response
  console.log("Files: ", res);
  alert("Upload Completed");
}}
onUploadBegin={(name) => {
  console.log("Started upload at: ", Date.now().toString());
  console.time('uploadbegin')
}}
onUploadProgress={(progress) => {
  console.timeEnd('uploadbegin')
  console.log("Progressed at: ", Date.now().toString(), ":", progress);
}}
```

**Comment by markflorkowski** [comment]
> I noticed that onUploadProgress not always being called with 0.

Thats... concerning. Just for that, I think it is worthwhile to do this.  

The delays aren't a huge concern for me, since in the case of small files, by the time the event hits, the file is probably uploaded already, and for larger files, a difference of 1 second is not really a big deal IMO.

**Comment by Mr0Bread** [comment]
Will be adding adjustments for SolidJS package then and docs

**Comment by Mr0Bread** [comment]
@markflorkowski @juliusmarminge Added changes for SolidJS package and also appended docs

Many changes from `pnpm format` btw

File: examples/appdir/src/app/page.tsx [hunk 1]
```diff
@@ -45,6 +45,9 @@ export default function Home() {
           //   },
           // }}
           className="ut-button:bg-red-400 ut-allowed-content:text-lg ut-button:ut-readying:bg-orange-300 rounded border p-3"
+          onUploadBegin={(name) => {
+            console.log("Uploading: ", name);
```
Comment (markflorkowski) [inline examples/appdir/src/app/page.tsx:49]: I would maybe add "onUploadBegin:" to the log here, just so anything in the console is more clear where it is coming from.

File: packages/react/src/components/dropzone.tsx [hunk 2]
```diff
@@ -57,17 +57,17 @@ export function UploadDropzone<TRouter extends FileRouter>(
   const $props = props as unknown as UploadDropzoneProps<TRouter> & {
     // props not exposed on public type
     // Allow to set internal state for testing
-    __internal_state?: "readying" | "ready" | "uploading",
+    __internal_state?: "readying" | "ready" | "uploading";
```
Comment (markflorkowski) [inline packages/react/src/components/dropzone.tsx:60]: how did this work before? 🤔

**Review by markflorkowski — APPROVED** [review]

File: packages/react/src/components/dropzone.tsx (hunk 2 above)
Comment (juliusmarminge) [reply packages/react/src/components/dropzone.tsx:60]: __internal_XXX are properties we added in the styling PR to be able to control the component state without user interaction to show the different states in the documentation

Comment (markflorkowski) [reply packages/react/src/components/dropzone.tsx:60]: No, I mean the commas in a type def. AFAIK that should throw terrible errors, no?

Comment (Mr0Bread) [reply packages/react/src/components/dropzone.tsx:60]: Not sure about the errors you are mentioning. Worked fine in this case

File: examples/appdir/src/app/page.tsx (hunk 1 above)
Comment (Mr0Bread) [reply examples/appdir/src/app/page.tsx:49]: Okay

File: packages/react/src/components/dropzone.tsx (hunk 2 above)
Comment (juliusmarminge) [reply packages/react/src/components/dropzone.tsx:60]: No, both commas and semicolons are valid in type objects

Comment (markflorkowski) [reply packages/react/src/components/dropzone.tsx:60]: Huh, TIL

File: examples/appdir/src/app/page.tsx (hunk 1 above)
Comment (Mr0Bread) [reply examples/appdir/src/app/page.tsx]: @markflorkowski done

Comment (juliusmarminge) [reply examples/appdir/src/app/page.tsx]: Did your latest commit mess up formatting?

We shoudl add a ci step for that 😅

Comment (Mr0Bread) [reply examples/appdir/src/app/page.tsx]: Sorry, just did run pnpm format and pushed

**Review by t3dotgg — APPROVED** [review]

**Pull request merged**